compiler.rb

```diff
@@ -70,31 +74,43 @@
   end
 end.parse!
 
-raise 'Option -p/--product is a required parameter' if product_name.nil?
+raise 'Cannt use -p/--products and -a/--all simultaneously' if product_names && all_products
+raise 'Either -p/--products OR -a/--all must be present' if product_names.nil? && !all_products
 raise 'Option -o/--output is a required parameter' if output_path.nil?
 raise 'Option -e/--engine is a required parameter' if provider_name.nil?
 
-product_yaml_path = File.join(product_name, 'api.yaml')
-raise "Product '#{product_name}' does not have an api.yaml file" \
-  unless File.exist?(product_yaml_path)
+if all_products
+  product_names = []
+  Dir["products/**/#{provider_name}.yaml"].each do |file_path|
+    product_names.push(File.dirname(file_path))
+  end
+
+  raise "No #{provider_name}.yaml files found. Check provider/engine name." if product_names.empty?
+end
+
+product_names.each do |product_name|
```

Comment:
And you're pretty sure there's no once-only global initialization that's provider-specific hiding someplace in here?  I am pretty sure, but not 100% sure.  I guess I wouldn't be surprised to see tiny trivial Terraform changes as a result of this PR.

Reply:
I'm not sure what you mean. Do you mean I should be moving this logic into the ruby code to take advantage of a global init or that this change is going to have unintended consequences due to global state? Or something else?